hngprojects/hng_boilerplate_nestjs
# feat: create testimonial

## What does the PR do?

This PR is about creating testimonials for logged in users. The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done

How should this be manually tested?

1. Login to a user account, and get the access token

2. Make a POST request to the `\api\testimonials` endpoint with the following payload:

```json
Authorization: Bearer <access_token>

{
  "name": "John Doe",
  "content": "The service is great. I love it. I hope to come back soon."
}
```

Successful Response

On a successful testimonial creation, the API should return a 201 CREATED status code.

```json
{
      "status": "success",
      "message": "Testimonial created successfully",
      "data": {
            "user_id": "a9f4b3b0-4b3b-4b3b-4b3b-4b3b4b3b4b3b",
            "name": "John Doe",
            "content": "The service is great. I love it. I hope to come back soon."
            "created_at":"2021-08-01T12:00:00Z"
      }
}
```

![image](https://github.com/user-attachments/assets/a0500b76-25f8-4742-a5fe-b586a9c03910)

Authentication Error:

When the current user is not properly authenticated to change the organization detail:

```json
{
  "status": "error",
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status_code": 401
}
```
![no token image](https://github.com/user-attachments/assets/171bb94f-6cd2-4acb-9643-026b85d0f791)

Invalid UserID Error

On Invalid error, if the user is not found with the provided id:

```json
{
  "status": "Error",
  "message": "user not found",
  "status_code": 404
}
```
![wrong user ID image](https://github.com/user-attachments/assets/6f9b6f4b-ff0d-4fa7-828f-aa61feff8440)

Validation:

If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages.

```json
{
  "status": "error",
  "message": ["name must be a string", "content must be a string"],
  "status_code": "400"
}
```
![validation_error_image](https://github.com/user-attachments/assets/77beadfc-2329-4583-a2f9-1134fa6b81d1)

Task

- [x] Create a POST endpoint PATCH /api/v1/testimonials/ to create testimonials for logged in users.
- [x] Implement server-side logic to handle successful and failed update request.
- [x] Ensure the user has the proper authentication to create testimonials.
- [x] Update the testimonials details in the database.
- [x] Write tests for creating testimonials.

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/206)
